src/components/templates/Camp/hooks/useKeyboardControls.ts

```diff
@@ -0,0 +1,219 @@
+  setIsHittingRoom: (value: boolean) => void;
+  takeSeat: (row: number | null, column: number | null) => void;
+
+  // Passed down from Camp component
+  venue: CampVenue;
+  selectedRoom?: CampRoomData;
+  setSelectedRoom: (room: CampRoomData | undefined) => void;
+}
+
+interface KeyboardControls {
+  roomEnter: (room: CampRoomData) => void;
+}
+
+export const useKeyboardControls: ReactHook<
+  UseKeyboardControlsProps,
+  KeyboardControls
+> = ({
+  user,
+  profile,
+
+  venue,
+  partygoersBySeat,
+
+  rows,
+  columns,
+  isHittingRoom,
+  setIsHittingRoom,
+  takeSeat,
+
+  selectedRoom,
+  setSelectedRoom,
+}) => {
+  const [keyDown, setKeyDown] = useState(false);
+
+  const venues = useSelector(orderedVenuesSelector);
+
+  const currentPosition = profile?.data?.[venue.id];
+
+  const downPress = useKeyPress("ArrowDown");
+  const upPress = useKeyPress("ArrowUp");
+  const leftPress = useKeyPress("ArrowLeft");
+  const rightPress = useKeyPress("ArrowRight");
+  const enterPress = useKeyPress("Enter");
+
+  const roomEnter = useCallback(
+    (room: CampRoomData) => {
+      const roomVenue = venues?.find((venue) =>
+        room.url.endsWith(`/${venue.id}`)
+      );
+      const venueRoom = roomVenue
+        ? { [roomVenue.name]: currentTimeInUnixEpoch }
+        : {};
+      room &&
+        user &&
+        enterRoom(
+          user,
+          {
+            [`${venue.name}/${room.title}`]: currentTimeInUnixEpoch,
+            ...venueRoom,
+          },
+          profile?.lastSeenIn
+        );
+    },
+    [profile, user, venue, venues]
+  );
+
+  // TODO: this is duplicated logic from handlePotentialRoomClick in Map component, so we can probably simplify
+  // TODO: do we want to filter this? Or just handle 1 room hit
+  const hitRoom = useCallback(
+    (row: number, column: number) => {
+      let isHitting = false;
+      venue.rooms.forEach((room: CampRoomData) => {
+        const rowPosition = (100 / rows) * row;
+        const colPosition = (100 / columns) * column;
+        const roomX = Math.round(room.x_percent);
+        const roomY = Math.round(room.y_percent);
+        const roomWidth = Math.round(room.width_percent);
+        const roomHeight = Math.round(room.height_percent);
+
+        if (
+          rowPosition >= roomY &&
+          rowPosition <= roomY + roomHeight &&
+          colPosition >= roomX &&
+          colPosition <= roomX + roomWidth
+        ) {
+          setSelectedRoom(room);
+          setIsHittingRoom(true);
+          isHitting = true;
+        } else {
+          if (isHittingRoom && selectedRoom === room) {
+            setSelectedRoom(undefined);
+            setIsHittingRoom(false);
+            isHitting = false;
+          }
+        }
+      });
+      return isHitting;
+    },
+    [
+      columns,
+      isHittingRoom,
+      venue.rooms,
+      rows,
+      selectedRoom,
+      setIsHittingRoom,
+      setSelectedRoom,
+    ]
+  );
+
+  useEffect(() => {
```

Comment:
[TODO] splitting this up into multiple `useEffect()`'s would allow us to downscope the amount of things that can cause it to break memo, not sure if that will make a huge difference here or not though.